File: Binary Search Question/find_minimum_in_rotated_sorted_array_ii_leetcode_154/find_minimum_in_rotated_sorted_array_ii_leetcode_154.java
// Problem Link   : https://leetcode.com/problems/find-minimum-in-rotated-sorted-array-ii/description/
// Company Tags   : None
// Youtube Link   : None
// Custom  Link   : None
// Similarities   : None

// Approach-1 (Normal)
// TC : average case O(logn) | worst case O(n)
// SC : O(1)
class Solution {
    public int findMin(int[] nums) {
        int s = 0;
        int e = nums.length - 1;

        while (s < e) {
            // why s < e ? not s <= e
            // e-- will converge to s == e
            // so, we need to check till s < e
            // when loop breaks... last condition to run will be e-- which guarantees.. s ==
            // e
            int mid = s + (e - s) / 2;

            if (nums[mid] > nums[e]) {
                s = mid + 1;
            } else if (nums[mid] < nums[e]) {
                e = mid;
            } else {
                e--;
            }
        }

        return nums[s];
    }

}

public class find_minimum_in_rotated_sorted_array_ii_leetcode_154 {
    public static void main(String[] args) {
        int[] nums = { 4, 5, 6, 1, 2, 3 };
        Solution sol = new Solution();
        System.out.println(sol.findMin(nums));
    }
}

File: Binary Search Question/find_minimum_in_rotated_sorted_array_ii_leetcode_154/find_minimum_in_rotated_sorted_array_ii_leetcode_154.py
# Problem Link   : https://leetcode.com/problems/find-minimum-in-rotated-sorted-array-ii/description/
# Company Tags   : None
# Youtube Link   : None
# Custom  Link   : None
# Similarities   : None

# Approach-1 (Normal)
# TC : average case O(logn) | worst case O(n)
# SC : O(1)
class Solution:
    def findMin(self, nums):
        s = 0
        e = len(nums) - 1

        while s < e:
            # why s < e ? not s <= e
            # e -= 1 will converge to s == e
            # so, we need to check till s < e
            # when loop breaks... last condition to run will be e -= 1 which gurantees.. s == e
            mid = s + (e - s) // 2

            if nums[mid] > nums[e]:
                s = mid + 1
            elif nums[mid] < nums[e]:
                e = mid
            else:
                e -= 1

        return nums[s]


if __name__ == "__main__":
    nums = [4, 5, 6, 1, 2, 3]
    sol = Solution()
    print(sol.findMin(nums))
